### ISSUE TITLE
shufflenetv2预训练的模型的问题

### ISSUE THREAD

**Issue by chenweifu2008** [ppogg/YOLOv5-Lite#3]
请问您有用shufflenetv2的预训练模型训练吗？
我要用pytorch官网的'shufflenetv2_x1.0' 模型的时候出现层的shape对应不上的问题：左边是官方的，右边是你的
10 : stage2.0.branch1.2.weight torch.Size([58, 24, 1, 1])    __   model.1.branch1.1.running_var torch.Size([24])
11 : stage2.0.branch1.3.weight torch.Size([58])    __     model.1.branch1.1.num_batches_tracked torch.Size([])
12 : stage2.0.branch1.3.bias torch.Size([58])    __       model.1.branch1.2.weight torch.Size([60, 24, 1, 1]

你的模型会出现多出一层‘model.1.branch1.1.num_batches_tracked’ 而且shape是[ ]空的，网上说这是新版pytorch的问题
但是主要问题是左边stage2.0.branch1.2  ([58, 24, 1, 1]) 和右边你的branch1.2.weight torch.Size([60, 24, 1, 1] 开始出现shape不匹配，接下去陆陆续续都开始shape不一样了

是您搭建网络出错了吗？

**Comment by ppogg** [comment]
有可能是你的预训练权重和我用的不一样,我把预训练权重的链接发给你:
    "shufflenetv2_0.5x": "https://download.pytorch.org/models/shufflenetv2_x0.5-f707e7126e.pth",  # noqa: E501
    "shufflenetv2_1.0x": "https://download.pytorch.org/models/shufflenetv2_x1-5666bf0f80.pth",  # noqa: E501
还有,需要保持yaml文件的一致:
# parameters
nc: 80  # number of classes
depth_multiple: 1.0  # model depth multiple
width_multiple: 1.0  # layer channel multiple

# anchors
anchors:
  - [ 4,5,  8,10,  13,16 ]  # P3/8
  - [ 23,29,  43,55,  73,105 ]  # P4/16
  - [ 146,217,  231,300,  335,433 ]  # P5/32

# custom backbone
backbone:
  # [from, number, module, args]
  [ [ -1, 1, conv_bn_relu_maxpool, [ 24 ] ],    # 0-P2/4
    [ -1, 1, Shuffle_Block, [ 116, 2 ] ], # 1-P3/8
    [ -1, 3, Shuffle_Block, [ 116, 1 ] ], # 2
    [ -1, 1, Shuffle_Block, [ 232, 2 ] ], # 3-P4/16
    [ -1, 7, Shuffle_Block, [ 232, 1 ] ], # 4
    [ -1, 1, Shuffle_Block, [ 464, 2 ] ], # 5-P5/32
    [ -1, 3, Shuffle_Block, [ 464, 1 ] ], # 6
  ]

# YOLOv5 head
head:
  [ [ -1, 1, Conv, [ 128, 1, 1 ] ],
    [ -1, 1, nn.Upsample, [ None, 2, 'nearest' ] ],
    [ [ -1, 4 ], 1, Concat, [ 1 ] ],  # cat backbone P4
    [ -1, 1, C3, [ 128, False ] ],  # 10

    [ -1, 1, Conv, [ 128, 1, 1 ] ],
    [ -1, 1, nn.Upsample, [ None, 2, 'nearest' ] ],
    [ [ -1, 2 ], 1, Concat, [ 1 ] ],  # cat backbone P3
    [ -1, 1, C3, [ 128, False ] ],  # 14 (P3/8-small)

    [ -1, 1, Conv, [ 128, 3, 2 ] ],
    [ [ -1, 11 ], 1, Concat, [ 1 ] ],  # cat head P4
    [ -1, 1, C3, [ 128, False ] ],  # 17 (P4/16-medium)

    [ -1, 1, Conv, [ 128, 3, 2 ] ],
    [ [ -1, 7 ], 1, Concat, [ 1 ] ],  # cat head P5
    [ -1, 1, C3, [ 128, False ] ],  # 20 (P5/32-large)

    [ [ 14, 17, 20 ], 1, Detect, [ nc, anchors ] ],  # Detect(P3, P4, P5)
  ]

**Comment by ppogg** [comment]
这是本人在安装上的适配,用的是以及使用了两年的红米手机:
fp16模型:
![mmexport1629451185220](https://user-images.githubusercontent.com/82716366/130212647-922d19be-67c2-4e1a-808b-ae1306c41d85.jpg)
这是量化后的int8模型:
![mmexport1629451003685](https://user-images.githubusercontent.com/82716366/130212693-1636c990-4b88-47a9-bfc4-3d1f885b574f.jpg)
其他模型和分辨率:
![qq_pic_merged_1629451239810](https://user-images.githubusercontent.com/82716366/130212804-190d35c4-a204-4d0f-b89d-051c419a240c.jpg)

**Comment by statham123** [comment]
题主，还有个输入对齐的问题，在你的代码中我发现没有做输入归一化到[0-1]和减均值除方差的操作，但是你给的pytorch官方model是需要这些步骤的，请指教

**Comment by ppogg** [comment]
你说的是部署在板子上cpp文件的归一化还是原模型文件的归一化呢？

**Comment by chenweifu2008** [comment]
> 有可能是你的预训练权重和我用的不一样,我把预训练权重的链接发给你:
> "shufflenetv2_0.5x": "https://download.pytorch.org/models/shufflenetv2_x0.5-f707e7126e.pth",
> "shufflenetv2_1.0x": "https://download.pytorch.org/models/shufflenetv2_x1-5666bf0f80.pth",
> 还有,需要保持yaml文件的一致:
> 

还是对不上
我用的是你给的这些
我是打印你训练的yolov5ss.pt 和shufflenetv2_1.0x 的weight，发现层的shape对应不上，名字
问题是yolo这两行代码很坑
state_dict = intersect_dicts(state_dict, model.state_dict(), exclude=exclude)  # intersect
model.load_state_dict(state_dict, strict=False)  # load
如果层数的名字对不上shape对不上他就自动跳过不加载了，也没报错了，让人以为加载了，如果把那个strict 改成True就会报不匹配

**Comment by ppogg** [comment]
我刚刚从这个仓库git clone了一下源码，data选用voc的person数据进行了测试，还是没发现对不齐的bug，会不会是你哪里出错了呢~

github: skipping check (not a git repository)
YOLOv5  2021-8-16 torch 1.8.0+cpu CPU

Namespace(adam=False, artifact_alias='latest', batch_size=4, bbox_interval=-1, bucket='', cache_images=False, cfg='models/yolov5ss-1.0.yaml', data='data/person.yaml', device='', entity=None, epochs=100, evolve=False, exist_ok=False, global_rank=-1, hyp='data/hyp.scratch.yaml', image_weights=False, img_size=[640, 640], label_smoothing=0.0, linear_lr=False, local_rank=-1, multi_scale=False, name='exp', noautoanchor=False, nosave=False, notest=False, project='runs/train', quad=False, rect=False, resume=False, save_dir='runs\\train\\exp', save_period=-1, single_cls=False, sync_bn=False, total_batch_size=4, upload_dataset=False, weights='weights/shufflenetv2_x1-5666bf0f80.pth', workers=8, world_size=1)
tensorboard: Start with 'tensorboard --logdir runs/train', view at http://localhost:6006/

hyperparameters: lr0=0.001, lrf=0.2, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=0.2, mixup=0.0
wandb: Install Weights & Biases for YOLOv5 logging with 'pip install wandb' (recommended)
Overriding model.yaml nc=80 with nc=1

                 from  n    params  module                                  arguments                     
  0                -1  1       696  models.common.conv_bn_relu_maxpool      [3, 24]                       
  1                -1  1      7764  models.common.Shuffle_Block             [24, 120, 2]                  
  2                -1  3     24300  models.common.Shuffle_Block             [120, 120, 1]                 
  3                -1  1     44588  models.common.Shuffle_Block             [120, 232, 2]                 
  4                -1  7    200564  models.common.Shuffle_Block             [232, 232, 1]                 
  5                -1  1    167968  models.common.Shuffle_Block             [232, 464, 2]                 
  6                -1  3    333384  models.common.Shuffle_Block             [464, 464, 1]                 
  7                -1  1     59648  models.common.Conv                      [464, 128, 1, 1]              
  8                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
  9           [-1, 4]  1         0  models.common.Concat                    [1]                           
 10                -1  1    104192  models.common.C3                        [360, 128, 1, False]          
 11                -1  1     16640  models.common.Conv                      [128, 128, 1, 1]              
 12                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 13           [-1, 2]  1         0  models.common.Concat                    [1]                           
 14                -1  1     89856  models.common.C3                        [248, 128, 1, False]          
 15                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 16          [-1, 11]  1         0  models.common.Concat                    [1]                           
 17                -1  1     90880  models.common.C3                        [256, 128, 1, False]          
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 19           [-1, 7]  1         0  models.common.Concat                    [1]                           
 20                -1  1     90880  models.common.C3                        [256, 128, 1, False]          
 21      [14, 17, 20]  1      6966  models.yolo.Detect                      [1, [[4, 5, 8, 10, 13, 16], [23, 29, 43, 55, 73, 105], [146, 217, 231, 300, 335, 433]], [128, 128, 128]]
Model Summary: 308 layers, 1533750 parameters, 1533750 gradients, 4.6 GFLOPS

image-size: 640
flops: 5.158330666666667
params: 1533750.0
Scaled weight_decay = 0.0005

Optimizer groups: 82 .bias, 82 conv.weight, 79 other
train: Scanning 'data\Person1K\labels\train.cache' images and labels... 1032 found, 0 missing, 33 empty, 0 corrupted: 100%|██████████| 1032/1032 [00:00<?, ?it/s]
val: Scanning 'data\Person1K\labels\val.cache' images and labels... 50 found, 0 missing, 0 empty, 0 corrupted: 100%|██████████| 50/50 [00:00<?, ?it/s]
Plotting labels... 

autoanchor: Analyzing anchors... anchors/target = 3.82, Best Possible Recall (BPR) = 1.0000
Image sizes 640 train, 640 test
Using 4 dataloader workers
Logging results to runs\train\exp
Starting training for 100 epochs...

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
      0/99        0G   0.09968   0.03064         0    0.1303         8       640:  36%|███▌      | 93/258 [03:36<07:05,  2.58s/it]

**Comment by chenweifu2008** [comment]
> 我刚刚从这个仓库git clone了一下源码，data选用voc的person数据进行了测试，还是没发现对不齐的bug，会不会是你哪里出错了呢~
> 

我用您的测试了，还是shape对不上，您网络输入的arguments 我看也没什么问题，但不知道为什么预训练模型的层的权值shape还是对不上，您的网络模型这层：model.1.branch1.2.weight：torch.Size([60, 24, 1, 1]) 预训练模型这层：stage2.0.branch1.2.weight： torch.Size([58, 24, 1, 1])
名字不一样可以改，但是shape不一样就没办法匹配了
